File: baekjoon/N1006.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class N1006 {
	static int[][][][][] memo;
	static int[][] map;
	static int n, w;

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int t = Integer.parseInt(br.readLine());
		for (int i = 0; i < t; i++) {
			StringTokenizer st = new StringTokenizer(br.readLine());
			n = Integer.parseInt(st.nextToken());
			w = Integer.parseInt(st.nextToken());
			map = new int[n][2];
			for (int j = 0; j < 2; j++) {
				st = new StringTokenizer(br.readLine());
				for (int k = 0; k < n; k++) {
					map[k][j] = Integer.parseInt(st.nextToken());
				}
			}
			memo = new int[n][2][2][2][2];
			int min;
			if (n == 1) {
				if (map[0][0] + map[0][1] <= w) {
					min = 1;
				} else {
					min = 2;
				}
			} else {
				min = Math.min(Math.min(dp(1, 1, 1, 0, 0), dp(1, 0, 0, 1, 1)),
						Math.min(dp(1, 1, 0, 0, 1), dp(1, 0, 1, 1, 0))) + 2;
				if (map[0][0] + map[0][1] <= w) {
					min = Math.min(min, dp(1, 0, 0, 0, 0) + 1);
				}
			}
			System.out.println(min);
		}
	}

	static int dp(int a, int b, int c, int d, int e) {
		if (memo[a][b][c][d][e] > 0) {
			return memo[a][b][c][d][e];
		}
		int ans = 2;
		if (a == n - 1) {
			if ((b == 1 && map[a - 1][0] + map[a][0] <= w) || (d == 1 && map[0][0] + map[a][0] <= w)) {
				ans--;
			}
			if ((c == 1 && map[a - 1][1] + map[a][1] <= w) || (e == 1 && map[0][1] + map[a][1] <= w)) {
				ans--;
			}
			if (ans == 2 && map[a][0] + map[a][1] <= w) {
				ans--;
			}
			return memo[a][b][c][d][e] = ans;
		}
		int min = Integer.MAX_VALUE;
		if (b == 1 && map[a - 1][0] + map[a][0] <= w) {
			if (c == 1 && map[a - 1][1] + map[a][1] <= w) {
				min = dp(a + 1, 0, 0, d, e);
			}
			min = Math.min(min, dp(a + 1, 0, 1, d, e) + 1);
		}
		if (c == 1 && map[a - 1][1] + map[a][1] <= w) {
			min = Math.min(min, dp(a + 1, 1, 0, d, e) + 1);
		}
		if (map[a][0] + map[a][1] <= w) {
			min = Math.min(min, dp(a + 1, 0, 0, d, e) + 1);
		}
		min = Math.min(min, dp(a + 1, 1, 1, d, e) + 2);
		return memo[a][b][c][d][e] = min;
	}
}

File: baekjoon/N1006.cpp
#include <stdio.h>
int en[10000][2];
int memo[10000][2][2][2][2];
int n, w;

int dp(int a, int b, int c, int d, int e)
{
	if (memo[a][b][c][d][e] != -1)
		return memo[a][b][c][d][e];
	if (a == n - 1)
	{
		memo[a][b][c][d][e] = 2;
		if (b + d > 0)
			memo[a][b][c][d][e]--;
		if (c + e > 0)
			memo[a][b][c][d][e]--;
		if (memo[a][b][c][d][e] == 2 && en[a][0] + en[a][1] <= w)
			memo[a][b][c][d][e] = 1;
		return memo[a][b][c][d][e];
	}
	memo[a][b][c][d][e] = 2 - b - c + dp(a + 1, 0, 0, d, e);
	if (b + c == 0 && en[a][0] + en[a][1] <= w)
		memo[a][b][c][d][e]--;
	if (en[a][0] + en[a + 1][0] <= w)
	{
		if (en[a][1] + en[a + 1][1] <= w)
			memo[a][b][c][d][e] = memo[a][b][c][d][e] < dp(a + 1, 1, 1, d, e) + 2 ? memo[a][b][c][d][e] : dp(a + 1, 1, 1, d, e) + 2;
		memo[a][b][c][d][e] = memo[a][b][c][d][e] < dp(a + 1, 1, 0, d, e) + 2 - c ? memo[a][b][c][d][e] : dp(a + 1, 1, 0, d, e) + 2 - c;
	}
	if (en[a][1] + en[a + 1][1] <= w)
		memo[a][b][c][d][e] = memo[a][b][c][d][e] < dp(a + 1, 0, 1, d, e) + 2 - b ? memo[a][b][c][d][e] : dp(a + 1, 0, 1, d, e) + 2 - b;
	return memo[a][b][c][d][e];
}

int main()
{
	int t;
	scanf("%d", &t);
	while (t--)
	{
		scanf("%d %d", &n, &w);
		for (int i = 0; i < n; i++)
			for (int j = 0; j < 16; j++)
				memo[i][j / 8][(j % 8) / 4][(j % 4) / 2][j % 2] = -1;
		for (int i = 0; i < 2; i++)
			for (int j = 0; j < n; j++)
				scanf("%d", &en[j][i]);
		int min = dp(0, 0, 0, 0, 0);
		if (en[0][0] + en[n - 1][0] <= w)
		{
			if (en[0][1] + en[n - 1][1] <= w)
				min = min < dp(0, 1, 1, 1, 1) + 2 ? min : dp(0, 1, 1, 1, 1) + 2;
			min = min < dp(0, 1, 0, 1, 0) + 1 ? min : dp(0, 1, 0, 1, 0) + 1;
		}
		if (en[0][1] + en[n - 1][1] <= w)
			min = min < dp(0, 0, 1, 0, 1) + 1 ? min : dp(0, 0, 1, 0, 1) + 1;
		printf("%d\n", min);
	}
}
